zeebe/broker/src/main/java/io/camunda/zeebe/broker/system/partitions/impl/MigrationSnapshotDirector.java

```diff
@@ -0,0 +1,155 @@
+import io.camunda.zeebe.util.health.HealthMonitorable;
+import io.camunda.zeebe.util.health.HealthReport;
+import java.time.Duration;
+import java.time.Instant;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentMap;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class MigrationSnapshotDirector implements HealthMonitorable, CloseableSilently {
+  public static final String COMPONENT_NAME = "MigrationSnapshotDirector";
+  private static final Logger LOG = LoggerFactory.getLogger(MigrationSnapshotDirector.class);
+
+  private volatile boolean snapshotTaken = false;
+  private ScheduledTimer runningSnapshot = null;
+  private final ConcurrentMap<FailureListener, Boolean> listeners = new ConcurrentHashMap<>();
+  private final AsyncSnapshotDirector snapshotDirector;
+
+  private volatile HealthReport healthReport;
+  private final boolean areMigrationsPerformed;
+  private final Duration scheduleDelay;
+  private final ConcurrencyControl control;
+  private final HealthMonitor healthMonitor;
+
+  public MigrationSnapshotDirector(
+      final AsyncSnapshotDirector asyncSnapshotDirector,
+      final boolean areMigrationsPerformed,
+      final ConcurrencyControl control,
+      final Duration scheduleDelay,
+      final HealthMonitor healthMonitor) {
+    snapshotDirector = asyncSnapshotDirector;
+    this.areMigrationsPerformed = areMigrationsPerformed;
+    this.control = control;
+    this.scheduleDelay = scheduleDelay;
+    this.healthMonitor = healthMonitor;
+    if (areMigrationsPerformed) {
+      healthReport = snapshotNotTaken();
+      scheduleSnapshot();
+    } else {
+      healthReport = HealthReport.healthy(this);
+    }
+    healthMonitor.registerComponent(this);
+    LOG.debug("Initialized migration snapshot director with status {}", healthReport);
+  }
+
+  @Override
+  public void close() {
+    cancelScheduledSnapshot();
+    healthMonitor.removeComponent(this);
+  }
+
+  public void scheduleSnapshot() {
+    if (areMigrationsPerformed && !snapshotTaken) {
+      LOG.debug("Scheduling snapshot for migration.");
+      forceSnapshotUntilSuccessful();
+    }
+  }
+
+  public void cancelScheduledSnapshot() {
+    if (runningSnapshot != null) {
+      runningSnapshot.cancel();
+      runningSnapshot = null;
+    }
+  }
+
+  @VisibleForTesting
+  public boolean isSnapshotTaken() {
+    return snapshotTaken;
+  }
+
+  @Override
+  public String componentName() {
+    return COMPONENT_NAME;
+  }
+
+  @Override
+  public HealthReport getHealthReport() {
+    return healthReport;
+  }
+
+  @Override
+  public void addFailureListener(final FailureListener failureListener) {
+    listeners.put(failureListener, Boolean.TRUE);
+  }
+
+  @Override
+  public void removeFailureListener(final FailureListener failureListener) {
+    listeners.remove(failureListener);
+  }
+
+  private void forceSnapshotUntilSuccessful() {
+    if (!snapshotTaken && runningSnapshot == null) {
```

Comment:
:question: This will loop forever? What if I have an unrecoverable error?

Reply:
Can we base it on the error itself? Some errors are unrecoverable, no? So maybe instead we just retry I/O exceptions only? Or can we look at the snapshot director and surface which errors are retry-able, and only retry those?